### PR TITLE
STORM-1040. SQL support for Storm.

### PR DESCRIPTION
**Pull request by haohui**
This PR rebases all the changes in the STORM-1040 branch towards the latest master.

### CODE REVIEW

**Comment by harshach** [comment]
+1. Great work @haohui to pull it off  :). Looking forward to get this into master branch. I've ran the tests with java 7 and java 8 all the tests are passing . Travis build failures not relevant to this PR.

**Comment by vesense** [comment]
+1 Great work

**Comment by darionyaphet** [comment]
@haohui Thank you for you Great Work :+1:  I want to know which SQL pattern had support ?  

such as count , distinct and sum ?

**Comment by hmcl** [comment]
+1. Nice job.

File: external/sql/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+CREATE EXTERNAL TABLE FOO (ID INT PRIMARY KEY) LOCATION 'kafka://localhost:2181/brokers?topic=test' TBLPROPERTIES '{"producer":{"bootstrap.servers":"localhost:9092","acks":"1","key.serializer":"storm.kafka.IntSerializer","value.serializer":"storm.kafka.ByteBufferSerializer"}}'
+```
+
+The syntax of `CREATE EXTERNAL TABLE` closely follows the one defined in [Hive Data Definition Language](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL).
+
+## Plugging in External Data Sources
+
+Users plug in external data sources through implementing the `ISqlTridentDataSource` interface and registers them using the mechanisms of Java's service loader. The external data source will be chosen based on the scheme of the URI of the tables. Please refer to the implementation of `storm-sql-kafka` for more details.
+
+It is also important to note that users also need to provide the dependency of the external data sources in the `extlib` directory. Otherwise the topology will fail to run because of `ClassNotFoundException`.
```
Comment (harshach) [inline external/sql/README.md]: can you highlight this part as a limitation section and file relevant JIRAs and include them in the README as well.

**Comment by haohui** [comment]
For some reasons I can't reopen this PR now. Created #919 as a duplicate for this PR.

Thanks everybody for the comments. I have addressed them in the latest PR.

**Comment by haohui** [comment]
Currently the implementation does not support any aggregation (i.e., count, sum, max) yet. It will be added once StormSQL supports windowing.

**Pull request closed without merging**